### ISSUE TITLE
Spellcheck custom language support

### ISSUE THREAD

**Issue by grissius** [typora/typora-issues#457]
It would be nice if you could use OpenOffice dictionaries and add custom languages. Abricotine md editor does it for instance.

**Comment by Halipster** [comment]
+1 would be super nice

**Comment by michalfita** [comment]
I liked Typora at first sight. I has features I miss in Evernote for article editions. But spell checking :man_facepalming:

**Comment by Rastagong** [comment]
This is the only feature missing for me too. I like to use Markdown for actual writing, but since I'm not a native English speaker, I need to be able to go back and forth between multiple languages. I'll have to get back to Markdown Pad 2 in the meantime…

Just wanted to add something: it seems that the dictionary used by Typora isn't actually English US at all times, but the system language dictionary? French is my system language, and a French dictionary is used when I enable spell check. 
(This is actually not helping in my case since I need to write in GB English.)

**Comment by abnerlee** [comment]
>  it seems that the dictionary used by Typora isn't actually English US at all times, but the system language dictionary? French is my system language, and a French dictionary is used when I enable spell check.
@Rastagong What's your OS? for Windows/Linux version, US-English will be used.

**Comment by Rastagong** [comment]
Windows 8.1, and French is my system language.

It's really weird because the bug seems to appear only after a restart of Typora. Here's what I did:

---

1. Before seeing your reply here, I had uninstalled Typora, thinking it would not do.
2. When I saw your reply, I reinstalled it, opened a document, but the bug wasn't present anymore. Spell check was working correctly with US English.
3. I disabled spell check, enabled it again. Spell check was still working correctly.
4. I restarted Typora, and there it happened again, it started to spell check in French (document below is fiction written in English, but suggested words are French): 
<img width="857" alt="typora spellcheck bug" src="https://user-images.githubusercontent.com/720624/30166528-e1ae8cb4-93e3-11e7-9341-bc7a585ad932.png">
5. I've done a couple more restarts (and enabled/disabled the spell check setting), but the problem persists.  

---

I'm not sure where it comes from. 😕   
Thanks for looking into it!

**Comment by Rastagong** [comment]
Oh, I did what I should have done before: just rebooted my laptop, and now US English is indeed correctly in use, no matter how many times I restart the app.

Sorry for the false warning! I'm not sure what caused this.

**Comment by abnerlee** [comment]
This is a bug anyway, we will keep an eye on this.

**Comment by belenos** [comment]
Typora uses the same OpenOffice.org Hunspell en_US dictionary (/usr/share/typora/resources/app/node_modules/spellchecker/vendor/hunspell_dictionaries) as Abricotine — which offers German, Spanish, Italian and Brazilian Portuguese as well. I believe the libraries are compatibles, right?

**Comment by fedriz** [comment]
I've done the italian dictionary for Abricotine (accepted pull request), can I add it Typora?

**Comment by abnerlee** [comment]
Based on doc for https://github.com/atom/node-spellchecker, it use Hunspell, or the Windows 8 Spell Check API, so Hunspell dict may only work on Windows 7 and Linux

**Comment by ArtanH** [comment]
For Windows 8 and newer, you need to edit line 29 of the **spellchecker.js** file which is located in
> C:\Program Files\Typora\resources\app\node_modules\spellchecker\lib\
```javascript
  var lang = process.env.LANG;
  lang = lang ? lang.split('.')[0] : '**en_US**';
  defaultSpellcheck = new Spellchecker();
```
Replace **en_US** with your language code. I use _de_DE_ for the german spell check.

**Comment by grissius** [comment]
@ArtanH 
![image](https://user-images.githubusercontent.com/24388536/38600622-1e78e2a4-3d65-11e8-8325-43488e145168.png)

**Comment by obiwankennedy** [comment]
I would like such a feature.

For the moment, I did the modification @ArtanH described but on linux.

The file to amend can be found here `/usr/share/typora/resources/app/node_modules/spellchecker/lib` on linux (ubuntu)

**Comment by nick-jansen** [comment]
@ArtanH great! But seems lik this isn't working for Dutch spellcheck...

**Comment by PalmtopTiger** [comment]
@nick-jansen try to change `'en-US'` to your language code (with a hyphen) in `C:\Program Files\Typora\resources\app\app\window\frame.js`.

**Comment by nick-jansen** [comment]
@PalmtopTiger thanks! Doesn't seem to work for switching to nl-NL on an English Windows 10 install.

**Comment by luroc** [comment]
@nick-jansen here is what I had to do to make it work (on Linux but could be similar on Windows): `Typora\resources\app\node_modules\spellchecker\vendor\hunspell_dictionaries ` had only en_US dictionaries for me, so I had to add the dictionary I wanted to use there.
I used the UTF-8 dictionary from [github.com/wooorm/dictionaries](https://github.com/wooorm/dictionaries) and placed them in the folder as ` de_DE.dic ` and ` de_DE.aff `. 
I ended up not changing  `29    lang = lang ? lang.split('.')[0] : 'en_US';` to ` de_DE`, but 
`28    var lang = process.env.LANG;` to 
` 28    var lang = 'de_DE.UTF-8';`.
Line 28 defines ` lang ` as your system locale and line 29 only applies `en_US` as a fallback. So If you don't have a system locale set (don't know how windows handles this) or you want a dictionary that differs from your system locale this is the way to go.

**Comment by Jaaaarne** [comment]
Windows 8. System language is Russian. Spellcheck only works for US English. Russian words are ignored, i.e. not underlined under no circumstances. In other apps including web browsers (Firefox, Opera) spellcheck works for different languages.

I tried all solutions suggested here. I was able to get Typora to check Russian text by combining suggestions by ArtanH and PalmtopTiger (only worked when I combined them, standalone solutions didn't work). 

Now I don't have English spellcheck. 

Basically, as it is now, Typora is unfit to be used by non-Americans where spellcheck is concerned. The developer should try to find the way to support other languages. Personally, I could easily live without localised interface, but spellcheck is very important. Hopefully this will be addressed, since this is a beta. 

Othe than that, a great app.
